doc/libpmempool.3
.\"
.\" Copyright 2016, Intel Corporation
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\"    * Redistributions of source code must retain the above copyright
.\"       notice, this list of conditions and the following disclaimer.
.\"
.\"     * Redistributions in binary form must reproduce the above copyright
.\"       notice, this list of conditions and the following disclaimer in
.\"       the documentation and/or other materials provided with the
.\"       distribution.
.\"
.\"     * Neither the name of the copyright holder nor the names of its
.\"       contributors may be used to endorse or promote products derived
.\"       from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
.\" A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
.\" OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
.\" OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\"
.\" libpmempool.3 -- man page for libpmempool
.\"
.\" Format this man page with:
.\"	man -l libpmempool.3
.\" or
.\"	groff -man -Tascii libpmempool.3
.\"
.TH libpmempool 3 "pmempool API version 1.0.1" "NVM Library"
.SH NAME
libpmempool \- pool management library
.SH SYNOPSIS
.nf
.B #include <libpmempool.h>
.sp
.B cc -std=gnu99 ... -lpmempool -lpmem
.sp
.B Health check functions:
.sp
.BI "PMEMpoolcheck *pmempool_check_init(struct pmempool_check_args *" args ,
.BI "    size_t " args_size ");"
.BI "struct pmempool_check_status *pmempool_check(PMEMpoolcheck *" ppc );
.BI "enum pmempool_check_result pmempool_check_end(PMEMpoolcheck *" ppc );
.sp
.B Library API versioning:
.sp
.BI "const char *pmempool_check_version("
.BI "    unsigned " major_required ,
.BI "    unsigned " minor_required );
.sp
.B Error handling:
.sp
.B const char *pmempool_errormsg(void);
.fi
.sp
.SH DESCRIPTION
.PP
.B libpmempool
provides a set of utilities for off-line analysis and manipulation of a
.IR pool .
By
.I pool
in this manpage we mean pmemobj pool, pmemblk pool, pmemlog pool or BTT layout,
independent of the underlying storage.
Some of
.B libpmempool
functions are required to work without any impact on processed
.I pool
but some of them may create a new or modify an existing one.
.PP
.B libpmempool
is for applications that need high reliability or built-in troubleshooting.
It may be useful for testing and debugging purposes also.
.PP
Currently
.B libpmempool
implements only consistency check and repair functions.
.sp
.SH POOL CHECKING FUNCTIONS
.PP
To perform check provided by
.BR libpmempool ", a"
.I check context
must be first initialized using
.BR pmempool_check_init ()
function described in this section. Once initialized
.I check context
is represented by an opaque handle, of type
.IR "PMEMpoolcheck *" ,
which is passed to all of the other functions described in this section.
.PP
To execute check
.BR pmempool_check ()
must be called iteratively. Each call resumes check till new status will be
generated. Each status is represented by
.B struct pmempool_check_status *
structure. It may carry various types of messages described in this section.
.PP
When check is completed
.BR pmempool_check ()
returns NULL pointer.
Check must be finalized using
.BR pmempool_check_end ().
It returns
.B enum pmempool_check_result
describing result of the whole check.
.PP
.nf
.BI "PMEMpoolcheck *pmempool_check_init(struct pmempool_check_args *" args ,
.BI "    size_t " args_size ");"
.fi
.IP
The
.BR pmempool_check_init ()
initializes
.IR "check context" ". " args
describes parameters of the
.IR "check context" .
.I args_size
should be equal to the size of the
.BR "struct pmempool_check_args" .
.B struct pmempool_check_args
is defined as follows:
.IP
.nf
struct pmempool_check_args {
    const char *path;                   /* path to the pool to check */
    const char *backup_path;            /* optional backup path */
    enum pmempool_pool_type pool_type;  /* type of the pool */
    int flags;                          /* parameters */
};
.fi
.IP
.I flags
accepts any combination of the following values (ORed):
.IP
.BR "" * " PMEMPOOL_CHECK_REPAIR"
- perform repairs
.IP
.BR "" * " PMEMPOOL_CHECK_DRY_RUN"
- emulate repairs
.IP
.BR "" * " PMEMPOOL_CHECK_ADVANCED"
- perform hazardous repairs
.IP
.BR "" * " PMEMPOOL_CHECK_ALWAYS_YES"
- do not ask before repairs
.IP
.BR "" * " PMEMPOOL_CHECK_VERBOSE"
- generate info statuses
.IP
.BR "" * " PMEMPOOL_CHECK_FORMAT_STR"
- generate string format statuses
.IP
If provided parameters are invalid or initialization process fails
.BR pmempool_check_init ()
returns NULL and sets errno appropriately.
.I pool_type
has to match type of the
.I pool
being processed.
You can turn on pool type detection by setting
.I pool_type
to
.B PMEMPOOL_POOL_TYPE_DETECT.
Pool type detection fail ends check.
.IP
.I backup_path
can either be:
.IP
* NULL. It indicates no backup will be performed.
.IP
* a non existing file. It is valid only in case
.I path
is a single file
.IR pool .
It indicates a
.I backup_path
file will be created and backup will be performed.
.IP
* an existing
.I poolset
file of the same structure (the same number of parts with exactly the same size)
as the source
.IR poolset .
It is valid only in case
.IR path " is a " poolset .
It indicates backup will be performed in a form described by the
.IR backup_path " " poolset .
.IP
This is an example of a
.I check context
initialization:
.IP
.nf
struct pmempool_check_args args = {
    .path           = "/path/to/blk.pool",
    .backup_path    = NULL,
    .pool_type      = PMEMPOOL_POOL_TYPE_BLK,
    .flags          = PMEMPOOL_CHECK_REPAIR | PMEMPOOL_CHECK_DRY_RUN |
        PMEMPOOL_CHECK_VERBOSE | PMEMPOOL_CHECK_FORMAT_STR
};

PMEMpoolcheck *ppc = pmempool_check_init(&args, sizeof(args));
.fi
.IP
The check will process a
.I pool
of type
.B PMEMPOOL_POOL_TYPE_BLK
located in the path /path/to/blk.pool. Before check it will not create a backup
of the
.IR "pool " ( "backup_path " == " " NULL).
If the check will find any issues it will try to perform repair steps
.BR "" ( PMEMPOOL_CHECK_REPAIR ),
but it will not make any changes to the
.I pool
.BR "" ( PMEMPOOL_CHECK_DRY_RUN )
and it will not perform any dangerous repair steps
.BR "" "(no " PMEMPOOL_CHECK_ADVANCED ).
The check will ask before performing any repair steps
.BR "" "(no " PMEMPOOL_CHECK_ALWAYS_YES ).
It will also generate detailed information about the check
.BR "" ( PMEMPOOL_CHECK_VERBOSE ).
.B PMEMPOOL_CHECK_FORMAT_STR
flag indicates string format statuses
.BR "" ( "struct pmempool_check_status" ).
Currently it is the only supported status format so this flag is required.
.PP
.BI "struct pmempool_check_status *pmempool_check(PMEMpoolcheck *" ppc );
.IP
The
.BR pmempool_check ()
function starts or resumes the check indicated by
.IR ppc .
When next status will be generated it pauses the check and returns a pointer to
the
.B struct pmempool_check_status
structure:
.IP
.nf
struct pmempool_check_status {
    enum pmempool_check_msg_type type;  /* type of the status */
    struct {
        const char *msg;                /* status message string */
        const char *answer;             /* answer to message if applicable */
    } str;
};
.fi
.IP
This structure can describe three types of statuses:
.IP
.BR "" * " PMEMPOOL_CHECK_MSG_TYPE_INFO"
- detailed information about the check. Generated only if a
.B PMEMPOOL_CHECK_VERBOSE
flag was set.
.IP
.BR "" * " PMEMPOOL_CHECK_MSG_TYPE_ERROR"
- encountered error
.IP
.BR "" * " PMEMPOOL_CHECK_MSG_TYPE_QUESTION"
- question. Generated only if an
.B PMEMPOOL_CHECK_ALWAYS_YES
flag was not set. It requires
.I answer
to be set to "yes" or "no" before continuing.
.IP
After calling
.BR pmempool_check ()
again the previously provided
.B struct pmempool_check_status *
pointer must be considered invalid. When the check completes
.BR pmempool_check ()
returns NULL pointer.
.PP
.BI "enum pmempool_check_result pmempool_check_end(PMEMpoolcheck *" ppc );
.IP
The
.BR pmempool_check_end ()
function finalizes the check and releases all related resources.
.I ppc
is not a valid pointer after calling
.BR pmempool_check_end ().
It returns
.B enum pmempool_check_result
summarizing result of the finalized check.
.BR pmempool_check_end ()
can return one of the following values:
.IP
.BR "" * " PMEMPOOL_CHECK_RESULT_CONSISTENT"
- the
.I pool
is consistent
.IP
.BR "" * " PMEMPOOL_CHECK_RESULT_NOT_CONSISTENT"
- the
.I pool
is not consistent
.IP
.BR "" * " PMEMPOOL_CHECK_RESULT_REPAIRED"
- the
.I pool
has issues but all repair steps completed succesfully
.IP
.BR "" * " PMEMPOOL_CHECK_RESULT_CANNOT_REPAIR"
- the
.I pool
has issues which can not be repaired
.IP
.BR "" * " PMEMPOOL_CHECK_RESULT_ERROR"
- the
.I pool
has errors or the check encountered issue
.SH LIBRARY API VERSIONING
.PP
This section describes how the library API is versioned, allowing applications
to work with an evolving API.
.PP
.BI "const char *pmempool_check_version("
.br
.BI "    unsigned " major_required ,
.br
.BI "    unsigned " minor_required );
.IP
The
.BR pmempool_check_version ()
function is used to see if the installed
.B libpmempool
supports the version of the library API required by an application.
The easiest way to do this for the application is to supply the compile-time
version information, supplied by defines in
.BR <libpmempool.h> ,
like this:
.IP
.nf
reason = pmempool_check_version(PMEMPOOL_MAJOR_VERSION, PMEMPOOL_MINOR_VERSION);
if (reason != NULL) {
    /* version check failed, reason string tells you why */
}
.fi
.IP
Any mismatch in the major version number is considered a failure,
but a library with a newer minor version number will pass this check since
increasing minor versions imply backwards compatibility.
.IP
An application can also check specifically for the existence of an interface
by checking for the version where that interface was introduced.
These versions are documented in this man page as follows: unless otherwise
specified, all interfaces described here are available in version 1.0 of the
library. Interfaces added after version 1.0 will contain the text introduced
in version x.y in the section of this manual describing the feature.
.IP
When the version check performed by
.BR pmempool_check_version ()
is successful, the return value is NULL. Otherwise the return value is a
static string describing the reason for failing the version check.
The string returned by
.BR pmempool_check_version ()
must not be modified or freed.
.SH DEBUGGING AND ERROR HANDLING
.PP
Two versions of
.B libpmempool
are typically available on a development system. The normal version, accessed
when a program is linked using the
.B -lpmempool
option, is optimized for performance. That version skips checks that impact
performance and exceptionally logs any trace information or performs any
run-time assertions. If an error is detected during the call to
.B libpmempool
function, an application may retrieve an error message describing the reason of
failure using the following function:
.PP
.BI "const char *pmempool_errormsg(void);
.IP
The
.BR pmempool_errormsg ()
function returns a pointer to a static buffer containing the last error message
logged for current thread. The error message may include description of the
corresponding error code (if errno was set), as returned by
.BR strerror (3).
The error message buffer is thread-local; errors encountered in one thread do
not affect its value in other threads. The buffer is never cleared by any
library function; its content is significant only when the return value of the
immediately preceding call to
.B libpmempool
function indicated an error, or if errno was set. The application must not
modify or free the error message string, but it may be modified by subsequent
calls to other library functions.
.PP
A second version of
.BR libpmempool ,
accessed when a program uses the libraries under
.BR /usr/lib/nvml_debug ,
contains run-time assertions and trace points. The typical way to access the
debug version is to set the environment variable
.BR "LD_LIBRARY_PATH " to " /usr/lib/nvml_debug " or " /usr/lib64/nvml_debug"
depending on where the debug libraries are installed on the system.
The trace points in the debug version of the library are enabled using the
environment variable
.BR PMEMPOOL_LOG_LEVEL ,
which can be set to the following values:
.IP 0
This is the default level when
.B PMEMPOOL_LOG_LEVEL
is not set. No log messages are emitted at this level.
.IP 1
Additional details on any errors detected are logged (in addition to returning
the errno-based errors as usual). The same information may be retrieved using
.BR pmempool_errormsg ().
.IP 2
A trace of basic operations is logged.
.IP 3
This level enables a very verbose amount of function call tracing in the library.
.IP 4
This level enables voluminous and fairly obscure tracing information that is
likely only useful to the libpmempool developers.
.PP
The environment variable
.B PMEMPOOL_LOG_FILE
specifies a file name where all logging information should be written.
If the last character in the name is "-", the PID of the current process will be
appended to the file name when the log file is created. If
.B PMEMPOOL_LOG_FILE
is not set, the logging output goes to stderr.
.PP
Setting the environment variable
.B PMEMPOOL_LOG_FILE
has no effect on the non-debug version of
.BR libpmempool .
.SH EXAMPLES
.PP
The following example illustrates how the
.B libpmempool
API is used.  The program detects the type and checks consistency of given pool.
If there are any issues detected, the pool is automatically repaired.
.IP
.\" run source through expand -4 before inserting...
.nf
#include <stddef.h>
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <libpmempool.h>

#define PATH "./pmem-fs/myfile"
#define CHECK_FLAGS (PMEMPOOL_CHECK_FORMAT_STR|PMEMPOOL_CHECK_REPAIR|\\
        PMEMPOOL_CHECK_VERBOSE)

int
main(int argc, char *argv[])
{
    PMEMpoolcheck *ppc;
    struct pmempool_check_status *status;
    enum pmempool_check_result ret;

    /* arguments for check */
    struct pmempool_check_args args = {
        .path       = PATH,
        .backup_path    = NULL,
        .pool_type  = PMEMPOOL_POOL_TYPE_DETECT,
        .flags      = CHECK_FLAGS
    };

    /* initialize check context */
    if ((ppc = pmempool_check_init(&args, sizeof(args))) == NULL) {
        perror("pmempool_check_init");
        exit(EXIT_FAILURE);
    }

    /* perform check and repair, answer 'yes' for each question */
    while ((status = pmempool_check(ppc)) != NULL) {
        switch (status->type) {
        case PMEMPOOL_CHECK_MSG_TYPE_ERROR:
            printf("%s\n", status->str.msg);
            break;
        case PMEMPOOL_CHECK_MSG_TYPE_INFO:
            printf("%s\n", status->str.msg);
            break;
        case PMEMPOOL_CHECK_MSG_TYPE_QUESTION:
            printf("%s\n", status->str.msg);
            status->str.answer = "yes";
            break;
        default:
            pmempool_check_end(ppc);
            exit(EXIT_FAILURE);
        }
    }

    /* finalize the check and get the result */
    ret = pmempool_check_end(ppc);
    switch (ret) {
        case PMEMPOOL_CHECK_RESULT_CONSISTENT:
        case PMEMPOOL_CHECK_RESULT_REPAIRED:
            return 0;
        default:
            return 1;
    }
}
.fi
.PP
See http://pmem.io/nvml/libpmempool for more examples
using the
.B libpmempool
API.
.SH ACKNOWLEDGEMENTS
.PP
.B libpmempool
builds on the persistent memory programming model recommended by the SNIA NVM
Programming Technical Work Group:
.IP
http://snia.org/nvmp
.SH "SEE ALSO"
.BR mmap (2),
.BR munmap (2),
.BR msync (2),
.BR strerror (3),
.BR libpmemobj (3),
.BR libpmemblk (3),
.BR libpmemlog (3),
.BR libpmem (3)
and
.BR http://pmem.io .


ChangeLog

Thu Jun 23 2016 Krzysztof Czurylo <krzysztof.czurylo@intel.com>

	* Version 1.1
	This NVML release introduces a new version of libpmemobj pool layout.
	Internal undo log structure has been modified to improve performance
	of pmemobj transactions.  Memory pools created with older versions
	of the libpmemobj library must be converted to the new format using
	"pmempool convert" command.  See pmempool-convert(1) for details.

	A new "libpmempool" library is available, providing support for
	off-line pool management and diagnostics.  Initially it provides only
	"check" and "repair" operations for log and blk memory pools, and for
	BTT devices.

	Other changes:
	- pmem: deprecate PCOMMIT
	- blk: match BTT Flog initialization with Linux NVDIMM BTT
	- pmem: defer pmem_is_pmem() initialization (#158)
	- obj: add TOID_TYPEOF macro

	Bug fixes:
	- doc: update description of valid file size units (#133)
	- pmempool: fix --version short option in man page (#135)
	- pmempool: print usage when running rm without arg (#136)
	- cpp: clarify polymorphism in persistent_ptr (#150)
	- obj: let the before flag be any non-zero value (#151)
	- obj: fix compare array pptr to nullptr (#152)
	- obj: cpp pool.get_root() fix (#156)
	- log/blk: set errno if replica section is specified (#161)
	- cpp: change exception message	(#163)
	- doc: remove duplicated words in man page (#164)
	- common: always append EXTRA_CFLAGS after our CFLAGS

	Experimental features:
	- Implementation of C++ bindings for libpmempobj is complete.
	  Web-based documentation for C++ API is available on http://pmem.io.
	  Note that C++ API is still considered experimental.  Do not use it
	  in production environments.
	- Porting NVML to Windows is in progress.  There are MS Visual Studio
	  solution/projects available, allowing to compile libpmem, libpmemlog,
	  libpmemblk and libpmemobj on Windows, but the libraries are not fully
	  functional and most of the test are not enabled yet.

Thu Apr 07 2016 Krzysztof Czurylo <krzysztof.czurylo@intel.com>

	* Version 1.0
	The API of six libraries (libpmem, libpmemblk, libpmemlog,
	libpmemobj, libvmem, libvmmalloc) is complete and stable.
	The on-media layout of persistent memory pools will be maintained
	from this point, and if changed it will be backward compatible.
	Man pages are all complete.
	This release has been validated to "Production quality".

	For the purpose of new features planned for next releases of NVML
	there have been some API modifications made:
		- pmem: pmem_map replaced with pmem_map_file
		- log/blk: 'off_t' substituted with 'long long'
		- obj: type numbers extended to 64-bit
		- obj: new entry points and macros added:
		  pmemobj_tx_errno, pmemobj_tx_lock, pmemobj_mutex_timedlock,
		  TX_ADD_DIRECT, TX_ADD_FIELD_DIRECT, TX_SET_DIRECT

	Other key changes since version 0.4 include:
		- common: updated/fixed installation scripts
		- common: eliminated dependency on libuuid
		- pmem: CPU features/ISA detection using CPUID
		- obj: improved error handling
		- obj: atomic allocation fails if constructor returns error
		- obj: multiple performance optimizations
		- obj: object store refactoring
		- obj: additional examples and benchmarks

	This release also introduces a prototype implementation
	of C++ bindings for libpmemobj.  Note that C++ API is still
	experimental and should not be used in production environments.

Fri Dec 04 2015 Krzysztof Czurylo <krzysztof.czurylo@intel.com>

	* Version 0.4
	This NVML version primarily focuses on improving code quality
	and reliability.  In addition to a couple of bug fixes,
	the changes include:
		- benchmarks for libpmemobj, libpmemblk and libvmem
		- additional pmemobj tests and examples
		- pool mapping address randomization
		- added pmempool "rm" command
		- eliminated libpmem dependency on libpthread
		- enabled extra warnings
		- minor performance improvements
	Man pages are all complete.
	This release is considered "Beta quality" by the team, having
	been thoroughly validated, including significant performance analysis.
	The pmempool command does not yet support "check" and "repair"
	operations for pmemobj type pools.

Sun Sep 13 2015 Andy Rudoff <andy.rudoff@intel.com>

	* Version 0.3
	NVML is now feature complete, adding support for:
		- pool sets
		- pmemobj local replication (active/passive)
		- experimental valgrind support
		- pmempool support for all pool types
	Man pages are all complete.
	This release is considered "Alpha quality" by the team, having
	gone through significant validation but only some performance
	analysis at this point.

Tue Jun 30 2015 Andy Rudoff <andy.rudoff@intel.com>

	* Version 0.2
	NVML now consists of six libraries:
		- libpmem                            (basic flushing, etc)
		- libpmemblk, libpmemlog, libpmemobj (transactions)
		- libvmem, libvmmalloc               (volatile use of pmem)
	The "pmempool" command is available for managing pmem files.
	Man pages for all the above are complete.
	The only things documented in man pages but not implemented are:
		- pmem sets (ability to spread a pool over a set of files)
		- replication (coming for libpmemobj)
	The pmempool command does not yet support pmemobj type pools.

Thu Sep 11 2014 Andy Rudoff <andy.rudoff@intel.com>

	* Version 0.1
	Initial development done in 0.1 builds


src/libpmempool/libpmempool.c
/*
 * Copyright 2016, Intel Corporation
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in
 *       the documentation and/or other materials provided with the
 *       distribution.
 *
 *     * Neither the name of the copyright holder nor the names of its
 *       contributors may be used to endorse or promote products derived
 *       from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * libpmempool.c -- entry points for libpmempool
 */

#include <stdlib.h>
#include <stdint.h>
#include <errno.h>
#include <sys/param.h>

#include "out.h"
#include "libpmempool.h"
#include "pmempool.h"
#include "pool.h"
#include "check.h"

/*
 * libpmempool_init -- load-time initialization for libpmempool
 *
 * Called automatically by the run-time loader.
 */
__attribute__((constructor))
static void
libpmempool_init(void)
{
	out_init(PMEMPOOL_LOG_PREFIX, PMEMPOOL_LOG_LEVEL_VAR,
		PMEMPOOL_LOG_FILE_VAR, PMEMPOOL_MAJOR_VERSION,
		PMEMPOOL_MINOR_VERSION);
	LOG(3, NULL);
	util_init();
}

/*
 * libpmempool_fini -- libpmempool cleanup routine
 *
 * Called automatically when the process terminates.
 */
__attribute__((destructor))
static void
libpmempool_fini(void)
{
	LOG(3, NULL);
	out_fini();
}

/*
 * pmempool_check_version -- see if library meets application version
 *	requirements
 */
const char *
pmempool_check_version(unsigned major_required, unsigned minor_required)
{
	LOG(3, "major_required %u minor_required %u",
			major_required, minor_required);

	if (major_required != PMEMPOOL_MAJOR_VERSION) {
		ERR("libpmempool major version mismatch (need %u, found %u)",
			major_required, PMEMPOOL_MAJOR_VERSION);
		return out_get_errormsg();
	}

	if (minor_required > PMEMPOOL_MINOR_VERSION) {
		ERR("libpmempool minor version mismatch (need %u, found %u)",
			minor_required, PMEMPOOL_MINOR_VERSION);
		return out_get_errormsg();
	}

	return NULL;
}

/*
 * pmempool_errormsg -- return last error message
 */
const char *
pmempool_errormsg(void)
{
	return out_get_errormsg();
}

/*
 * pmempool_ppc_set_default -- (internal) set default values of check context
 */
static void
pmempool_ppc_set_default(PMEMpoolcheck *ppc)
{
	const PMEMpoolcheck ppc_default = {
		.args		= {
			.path		= NULL,
			.backup_path	= NULL,
			.pool_type	= PMEMPOOL_POOL_TYPE_DETECT,
			.flags		= 0
		},
		.data		= NULL,
		.pool		= NULL,
		.result		= CHECK_RESULT_CONSISTENT,
	};
	*ppc = ppc_default;
}

/*
 * pmempool_check_init -- initialize check context
 */
PMEMpoolcheck *
pmempool_check_init(struct pmempool_check_args *args, size_t args_size)
{
	LOG(3, "path %s backup_path %s pool_type %u flags %x", args->path,
		args->backup_path, args->pool_type, args->flags);

	/*
	 * Currently one size of args structure is supported. The version of the
	 * pmempool_check_args structure can be distinguished based on provided
	 * args_size.
	 */
	if (args_size < sizeof(struct pmempool_check_args)) {
		ERR("provided args_size is not supported");
		errno = EINVAL;
		return NULL;
	}

	/*
	 * Dry run does not allow to made changes possibly performed during
	 * repair. Advanced allow to perform more complex repairs. Questions
	 * are ask only if repairs are made. So dry run, advanced and always_yes
	 * can be set only if repair is set.
	 */
	if (util_flag_isclr(args->flags, PMEMPOOL_CHECK_REPAIR) &&
			util_flag_isset(args->flags, PMEMPOOL_CHECK_DRY_RUN |
			PMEMPOOL_CHECK_ADVANCED | PMEMPOOL_CHECK_ALWAYS_YES)) {
		ERR("dry_run, advanced and always_yes are applicable only if "
			"repair is set");
		errno = EINVAL;
		return NULL;
	}

	/*
	 * dry run does not modify anything so performing backup is redundant
	 */
	if (util_flag_isset(args->flags, PMEMPOOL_CHECK_DRY_RUN) &&
			args->backup_path != NULL) {
		ERR("dry run does not allow to perform backup");
		errno = EINVAL;
		return NULL;
	}

	/*
	 * libpmempool uses str format of communication so it must be set
	 */
	if (util_flag_isclr(args->flags, PMEMPOOL_CHECK_FORMAT_STR)) {
		ERR("PMEMPOOL_CHECK_FORMAT_STR flag must be set");
		errno = EINVAL;
		return NULL;
	}

	PMEMpoolcheck *ppc = malloc(sizeof(*ppc));
	if (ppc == NULL) {
		ERR("!malloc");
		return NULL;
	}

	pmempool_ppc_set_default(ppc);
	memcpy(&ppc->args, args, sizeof(ppc->args));
	ppc->path = strdup(args->path);
	if (!ppc->path) {
		ERR("!strdup");
		goto error_path_malloc;
	}
	ppc->args.path = ppc->path;

	if (args->backup_path != NULL) {
		ppc->backup_path = strdup(args->backup_path);
		if (!ppc->args.backup_path) {
			ERR("!strdup");
			goto error_backup_path_malloc;
		}
		ppc->args.backup_path = ppc->backup_path;
	}

	if (check_init(ppc) != 0)
		goto error_check_init;

	return ppc;

error_check_init:
	/* in case errno not set by any of the used functions set its value */
	if (errno == 0)
		errno = EINVAL;

	free(ppc->backup_path);
error_backup_path_malloc:
	free(ppc->path);
error_path_malloc:
	free(ppc);
	return NULL;
}

/*
 * pmempool_check -- continue check till produce status to consume for caller
 */
struct pmempool_check_status *
pmempool_check(PMEMpoolcheck *ppc)
{
	LOG(3, NULL);
	ASSERTne(ppc, NULL);

	struct check_status *result;
	do {
		result = check_step(ppc);

		if (check_is_end(ppc->data) && result == NULL)
			return NULL;
	} while (result == NULL);

	return check_status_get(result);
}

/*
 * pmempool_check_end -- end check and release check context
 */
enum pmempool_check_result
pmempool_check_end(PMEMpoolcheck *ppc)
{
	LOG(3, NULL);
	enum check_result result = ppc->result;

	check_fini(ppc);
	free(ppc->path);
	free(ppc->backup_path);
	free(ppc);

	switch (result) {
		case CHECK_RESULT_CONSISTENT:
			return PMEMPOOL_CHECK_RESULT_CONSISTENT;

		case CHECK_RESULT_NOT_CONSISTENT:
			return PMEMPOOL_CHECK_RESULT_NOT_CONSISTENT;

		case CHECK_RESULT_REPAIRED:
			return PMEMPOOL_CHECK_RESULT_REPAIRED;

		case CHECK_RESULT_CANNOT_REPAIR:
			return PMEMPOOL_CHECK_RESULT_CANNOT_REPAIR;

		default:
			return PMEMPOOL_CHECK_RESULT_ERROR;
	}
}
